cmd/vicadmin/server.go

```diff
@@ -114,39 +130,158 @@ func (s *server) listenPort() int {
+	s.Authenticated(link, h.ServeHTTP)
+}
+
+func (s *server) Handle(link string, h http.Handler) {
+	s.mux.Handle(link, gorillacontext.ClearHandler(h))
+}
+
+// Enforces authentication on route `link` and runs `handler` on successful auth
+func (s *server) Authenticated(link string, handler func(http.ResponseWriter, *http.Request)) {
+	defer trace.End(trace.Begin(""))
+
+	authHandler := func(w http.ResponseWriter, r *http.Request) {
+		websession, err := store.Get(r, sessionCookieKey)
+		if err != nil {
+			log.Errorf("Got \"%s\" while looking up cookie store", err.Error())
+		}
+
+		if len(r.TLS.PeerCertificates) > 0 { // the user is authenticated by certificate at connection time
+			usersess := s.uss.Add("root", &rootConfig.Config)
+			websession, _ := store.Get(r, sessionCookieKey)
+			timeNow, _ := usersess.created.MarshalText()
+			websession.Values["created"] = string(timeNow)
+			websession.Values["username"] = "root"
+			websession.Save(r, w)
+			handler(w, r)
+			return
+		}
+
+		c := websession.Values["created"]
+		if c == nil { // no cookie, so redirect to login
+			log.Errorf("No authentication token: %+v", websession.Values)
+			http.Redirect(w, r, "/authentication", 302)
+			return
+		}
+
+		// parse the cookie creation time
+		created, _ := time.Parse(time.RFC3339, c.(string))
+
+		if time.Since(created) > sessionExpiration { // cookie exists but is expired
+			http.Redirect(w, r, "/authentication?expired", 302)
+			return
+		}
+
+		// if the date on the cookie was valid, then the user is authenticated
+		handler(w, r)
+	}
+	s.mux.Handle(link, gorillacontext.ClearHandler(http.HandlerFunc(authHandler)))
+}
+
+// renders the page for login and handles authorization requests
+func (s *server) loginPage(res http.ResponseWriter, req *http.Request) {
+	defer trace.End(trace.Begin(""))
+	ctx := context.Background()
+	if req.Method == "POST" {
+		// take the form data and use it to try to authenticate with vsphere
+
+		// create a userconfig
+		userconfig := session.Config{
+			Insecure:       false,
+			Thumbprint:     rootConfig.Thumbprint,
+			Keepalive:      rootConfig.Keepalive,
+			ClusterPath:    rootConfig.ClusterPath,
+			DatacenterPath: rootConfig.DatacenterPath,
+			DatastorePath:  rootConfig.DatastorePath,
+			HostPath:       rootConfig.Config.HostPath,
+			PoolPath:       rootConfig.PoolPath,
+		}
+		user := url.UserPassword(req.FormValue("username"), req.FormValue("password"))
+		serviceURL, err := soap.ParseURL(rootConfig.Service)
+		if err != nil {
+			log.Errorf("vSphere service URL was not a valid format; parsing returned error: %s", err)
+		}
+		serviceURL.User = user
+		userconfig.Service = serviceURL.String()
+
+		// check login
+		usersession, err := vSphereSessionGet(&userconfig)
+		if err != nil || usersession == nil {
+			// something went wrong or we could not authenticate
+			log.Warnf("User %s failed to authenticated at %s", user, time.Now())
+			http.Redirect(res, req, "/authentication?unauthorized", 302)
```

Comment:
Yeah you're right, I'll make sure I have `return` statements after all of my `Redirect`s